TL/CUDA: Alltoallv implementation

kingchc opened the pull request:
## What
Implement TL CUDA alltoallv

## Why ?
Missing feature

## How ?
This PR is based on existing the TL CUDA Alltoall implementation and has the following differences to support Alltoallv.
- use info_v to get correct buffer ptr, counts, displacements, etc...
- extend shared memory region to store additional info for Alltoallv, e.g., counts, displacements, for proxies to access if applicable

swx-jenkins3 commented:
Can one of the admins verify this patch?

kingchc commented:
cc @Sergei-Lebedev @vspetrov @bureddy

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,316 @@
+    data_len = ucc_dt_size(args->src.info_v.datatype) *
+               ucc_coll_args_get_total_count(args, args->src.info_v.counts,
+                                             UCC_TL_TEAM_SIZE(team));
+    status = ucc_tl_cuda_mem_info_get(args->src.info_v.buffer, data_len, team,
```
src and dst buffers are not always contig in alltoallv, so mem info get might fail here.
As a quick solution pls add check for UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER and UCC_COLL_ARGS_FLAG_CONTIG_DST_BUFFER and similarly to inplace flag return UCC_ERR_NOT_SUPPORTED

kingchc replied:
Added a check in the beginning of `ucc_tl_cuda_alltoallv_ce_init`.

Sergei-Lebedev commented on src/components/tl/cuda/tl_cuda.h:
```diff
@@ -151,12 +156,12 @@ typedef struct ucc_tl_cuda_task {
+            void                  *peer_map_addr_src[UCC_TL_CUDA_MAX_PEERS];
+            void                  *peer_map_addr_dst[UCC_TL_CUDA_MAX_PEERS];
+            void                  *copy_done;
         } alltoall_ce;
```
minor, i understand that alltoall_ce and alltoallv_ce algorithms will probably be identical, but anyway i think it's better to have a different struct for alltoallv

kingchc replied:
done

Sergei-Lebedev commented:
@kingchc thanks for the PR! overall looks good to me, let me do some testing internally

Sergei-Lebedev commented:
ok to test

kingchc commented on src/components/tl/cuda/tl_cuda.h:
```diff
@@ -113,19 +114,23 @@ typedef struct ucc_tl_cuda_sync_data {
+    ucc_tl_cuda_mem_info_t  mem_info_dst;
+    cudaEvent_t             ipc_event_local;
+    cudaIpcEventHandle_t    ev_handle;
+    ucc_tl_cuda_sync_data_t data[UCC_TL_CUDA_MAX_PEERS];
```
@Sergei-Lebedev - I have changed this from `data[1];` to `data[UCC_TL_CUDA_MAX_PEERS];` to avoid index-out-of-range error when running alltoallv (when perfoming cudaStreamWaitEvent). I'm surprised it works fine for alltoall, please let me know if this change looks okay to you.

kingchc replied:
this is fixed by moving around the data structure of `ucc_tl_cuda_sync_t` as this PR adds new fields like `src_cnts`.

vspetrov commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,319 @@
+    ucc_status_t        status;
+    size_t              data_len;
+
+    if (!(args->flags & (UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER |
```
i think the check is not correct:
1. you also need to check if args->mask & UCC_COLL_ARGS_FIELD_FLAGS is there (otherwise flags can have garbage)
2. you check would pass if either src or dst contig is set, while you need both

I would add helper macros to ucc_Coll_utils:
UCC_ARGS_SRC_CONTIG(args) (args.mask & UCC_COLL_ARGS_FIELD_FLAGS) && (args.flags & UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER)
similar for dst
and
for both:
UCC_COLL_ARGS_CONTIG_BUFFER(args) UCC_COLL_ARGS_SRC_CONTIG(args) && UCC_COLL_ARGS_DST_CONTIG(args)

vspetrov commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,319 @@
+    // copy a2av counts and displ. for proxies (if any) to access
+    if (team->topo->num_proxies > 0) {
+        ucc_coll_args_t *args = &TASK_ARGS(task);
+        // TODO: would team size match the size in args when using HIER CL?
```
yes, CL/HIER sets up counts/displs appropriately

vspetrov commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,319 @@
+    if (team->topo->num_proxies > 0) {
+        ucc_coll_args_t *args = &TASK_ARGS(task);
+        // TODO: would team size match the size in args when using HIER CL?
+        memcpy(sync->src_cnts, args->src.info_v.counts,
```
the problem with these memcpy is that counts/displs may be either 32 or 64 bits wide, so you can't use sizeof(ucc_count_t). Need to do smth like:
elem_size = UCC_COLL_ARGS_COUNT64(args) ? 8 : 4;
memcpy(..., elem_size * team_size)

vspetrov commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,319 @@
+    return status;
+}
+
+ucc_status_t ucc_tl_cuda_alltoallv_setup_test(ucc_tl_cuda_task_t *task)
```
this fn looks exactly same as alltoall_setup_test - can we unify them?

vspetrov reviewed (commented):
In general alltoallv start/progress looks very similar to alltoall. The only main difference (except for setup) is get_count/get_displ. Probably we could abstract them (adding fn pointer to task) and then re-use the same code for both algs. Similar thing is done by @Sergei-Lebedev in #437 (e.g. tl/cuda/allgatherv/allgatherv_ring.c, in tl_cuda_allgatherv_ring_init, look for ucc_tl_cuda_allgatherv_get_count)

kingchc commented:
@Sergei-Lebedev @vspetrov @bureddy - This PR is ready for review again, major changes.
- rebased on top of latest master
- unify alltoall and alltoallv similar to #427 
- addressed above comments

Please let me know if you have any further comments

Sergei-Lebedev commented:
CI error
```
/opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c: In function 'ucc_tl_cuda_alltoallv_ce_post_copies':
/opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c:122:33: error: unused variable 'args' [-Werror=unused-variable]
02:14:31       ucc_coll_args_t            *args = &TASK_ARGS(task);```

kingchc commented:
> CI error
> 
> ```
> /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c: In function 'ucc_tl_cuda_alltoallv_ce_post_copies':
> /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c:122:33: error: unused variable 'args' [-Werror=unused-variable]
> 02:14:31       ucc_coll_args_t            *args = &TASK_ARGS(task);```
> ```

Thanks for catching this. Fix pushed.

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,402 @@
+                                    sync->data[peer].ipc_event_remote, 0),
+                exit, status);
+        }
+        data_size = sdt_size * task->alltoallv_ce.get_count(
```
for alltoall, get_count function returns number of elements in dst buffer which might be different from number of elements in src buffer, so need to multiple by rdt_size

kingchc replied:
hmm...in the case of a remote process uses different datatype, I think both rdt_size and sdt_size could be wrong. I will take your suggestion below to save size and offset in bytes to solve this problem.

kingchc replied:
is it correct? If the src is rank itself (i.e., `peer==rank`), then it should be `sdt_size`. If src is from remote GPU, it should be remote's `sdt_size` which might not be the same as the local `sdt_size`/`rdt_size`, if I understand the semantic correctly. I like the idea you mentioned below of saving number of bytes instead elements to make things easier.

Sergei-Lebedev replied:
right, let's just save number of bytes in shm segment 😊

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,402 @@
+        }
+        data_size = sdt_size * task->alltoallv_ce.get_count(
+                                   task, task->alltoallv_ce.scnts, peer);
+        data_displ = sdt_size * task->alltoallv_ce.get_offset(
```
since you calculate displacement in remote buffer you need to use offset from shared memory i.e.
```
data_displ = sdt_size * task->alltoallv_ce.get_offset(task, peer_sync->src_displ, rank);
```

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,402 @@
+    if (ucc_unlikely(status != UCC_OK)) {
+        return status;
+    }
+
```
need to add task->super.flags |= UCC_COLL_TASK_FLAG_EXECUTOR to tell upper layer to create executor for regular post

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,402 @@
+    if (team->topo->num_proxies > 0 &&
+        UCC_COLL_TYPE_ALLTOALLV == args->coll_type) {
+        size_t elem_size = UCC_COLL_ARGS_COUNT64(args) ? 8 : 4;
+        memcpy(sync->src_cnts, args->src.info_v.counts,
```
it would be easier if we save offsets and counts in bytes rather than in elements, it will solver sdt_size and rdt_size problem and also enable support for the case when source datatype is not the same as recv datatype

kingchc commented:
Addressed Sergey's comments. Major changes:
- rebased
- cache A2Av sizes and displacement **in bytes** on SHM, e.g., `get_count`->`get_size`, to make remote access easier without explicitly handling about src/dst datatype

kingchc commented:
There seems to be some tests failed, but I could not see the details. @Sergei-Lebedev - plz let me know if you see any runtime issue, and I can fix them.

Sergei-Lebedev commented:
> There seems to be some tests failed, but I could not see the details. @Sergei-Lebedev - plz let me know if you see any runtime issue, and I can fix them.

it's compilation issue
```
00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c: In function 'ucc_tl_cuda_alltoallv_setup_start':
00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c:66:5: error: implicit declaration of function 'ucc_memory_bus_fence'; did you mean 'ucc_memory_cpu_fence'? [-Werror=implicit-function-declaration]
00:27:54       ucc_memory_bus_fence();
```
if https://github.com/openucx/ucc/pull/442 ucc_memory_bus_fence was removed, now it should be ucc_memory_cpu_store_fence();

kingchc commented:
> > There seems to be some tests failed, but I could not see the details. @Sergei-Lebedev - plz let me know if you see any runtime issue, and I can fix them.
> 
> it's compilation issue
> 
> ```
> 00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c: In function 'ucc_tl_cuda_alltoallv_setup_start':
> 00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c:66:5: error: implicit declaration of function 'ucc_memory_bus_fence'; did you mean 'ucc_memory_cpu_fence'? [-Werror=implicit-function-declaration]
> 00:27:54       ucc_memory_bus_fence();
> ```
> 
> if #442 ucc_memory_bus_fence was removed, now it should be ucc_memory_cpu_store_fence();

Thanks for pointing it out! looks like I tested a wrong version locally and did not catch this. Fixes pushed.

kingchc commented:
> > There seems to be some tests failed, but I could not see the details. @Sergei-Lebedev - plz let me know if you see any runtime issue, and I can fix them.
> 
> it's compilation issue
> 
> ```
> 00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c: In function 'ucc_tl_cuda_alltoallv_setup_start':
> 00:27:54  /opt/nvidia/src/ucc/src/components/tl/cuda/alltoallv/alltoallv_ce.c:66:5: error: implicit declaration of function 'ucc_memory_bus_fence'; did you mean 'ucc_memory_cpu_fence'? [-Werror=implicit-function-declaration]
> 00:27:54       ucc_memory_bus_fence();
> ```
> 
> if #442 ucc_memory_bus_fence was removed, now it should be ucc_memory_cpu_store_fence();

Thanks for pointing it out! looks like I tested a wrong version locally and did not catch this.

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,405 @@
+    switch (task->alltoallv_ce.stage) {
+    case ALLTOALL_CE_STAGE_SYNC:
+        if (ucc_tl_cuda_get_sync(task) != UCC_OK) {
+            task->super.super.status = UCC_INPROGRESS;
```
need to replace all task->super.super.status = something with task->super.status = something, because of this change https://github.com/openucx/ucc/pull/441

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,405 @@
+        task->alltoallv_ce.stage = ALLTOALL_CE_STAGE_SYNC;
+    }
+
+    ucc_tl_cuda_alltoallv_ce_progress(coll_task);
+    if (task->super.super.status == UCC_INPROGRESS) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    return ucc_task_complete(coll_task);
```
pls replace with, 
```suggestion
return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
```
ucc_progress_queue_enqueue will do all the necessary setup for task,

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,405 @@
+                                    sync->data[peer].ipc_event_remote, 0),
+                exit, status);
+        }
+        data_size =
+            task->alltoallv_ce.get_size(task, sync->alltoallv_ce.sbytes, peer);
+        data_displ = task->alltoallv_ce.get_offset(
+            task, sync->alltoallv_ce.sdispl_bytes, peer);
```
data_displ is not correct
```suggestion
        data_size =
            task->alltoallv_ce.get_size(task, peer_sync->alltoallv_ce.sbytes, rank);
        data_displ = task->alltoallv_ce.get_offset(
            task, peer_sync->alltoallv_ce.sdispl_bytes, rank);
```

kingchc commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,405 @@
+            task->alltoallv_ce.num_posted++;
+        }
+    }
+    status = ucc_ec_event_post(team->stream, task->alltoallv_ce.copy_done,
+                               UCC_EE_CUDA_STREAM);
```
@Sergei-Lebedev - I just realized this will result in calling `cudaEventRecord` and may cause deadlock for multi-thread case (e.g., PyTorch). It seems to me we don't really need this now since we introduce the executor. 

I plan to remove this and all `task->alltoallv_ce.copy_done`, please let me know if you think otherwise.

Sergei-Lebedev commented on src/components/tl/cuda/alltoallv/alltoallv_ce.c:
```diff
@@ -0,0 +1,389 @@
+    size_t              data_len;
+
+    if (!UCC_COLL_ARGS_CONTIG_BUFFER(args)) {
+        tl_error(UCC_TL_TEAM_LIB(team), "Do not support non-contiguous buffer");
```
I think it's better to use tl_debug here since we will fallback to other TL. I see a lot of errors in internal tests log but test passes anyway
```
[1649680880.572873]     alltoallv_ce.c:344  TL_CUDA ERROR Do not support non-contiguous buffer
```

kingchc replied:
I'm concerned that users might make wrong assumptions if we fall back silently. Perhaps we can just print the error the first time?

vspetrov replied:
@kingchc the behavior we are trying to keep in UCC is: be functional as much as possible (means, if smth is not available/not implement try to fallback to other components) and be "silent" by default unless fatal error (like NO_MEM). User can always check with proper LOG_LEVEL setting

Sergei-Lebedev reviewed (approved):
PR looks good to me, thanks @kingchc!

kingchc commented:
rebased and addressed review comments. (cc @vspetrov @Sergei-Lebedev)
